Sources/Plasma/Apps/plClient/Mac-Cocoa/PLSLoginController.mm

```diff
@@ -0,0 +1,183 @@
+containing parts covered by the terms of the Bink SDK EULA, 3ds Max EULA,
+PhysX SDK EULA, DirectX SDK EULA, OpenSSL and SSLeay licenses, IJG
+JPEG Library README, Windows Media SDK EULA, or QuickTime SDK EULA, the
+licensors of this Program grant you additional
+permission to convey the resulting work. Corresponding Source for a
+non-source form of such a combination shall include the source code for
+the parts of OpenSSL and IJG JPEG Library used as well as that of the covered
+work.
+
+You can contact Cyan Worlds, Inc. by email legal@cyan.com
+ or by snail mail at:
+      Cyan Worlds, Inc.
+      14617 N Newport Hwy
+      Mead, WA   99021
+
+*==LICENSE==*/
+
+#include <regex>
+
+#include "PLSLoginController.h"
+
+#include "hsEndian.h"
+#include "NSString+StringTheory.h"
+#include "pfPasswordStore/pfPasswordStore.h"
+#include "PLSServerStatus.h"
+#include "plNetGameLib/plNetGameLib.h"
+#include "plProduct.h"
+#include "plNetClient/plNetClientMgr.h"
+#include "pnEncryption/plChallengeHash.h"
+
+#define FAKE_PASS_STRING @"********"
+
+static NSOperationQueue* _loginQueue = nil;
+
+@implementation PLSLoginController
+
++ (void)initialize
+{
+    _loginQueue = [NSOperationQueue new];
+    _loginQueue.maxConcurrentOperationCount = 1;
+}
+
++ (void)attemptLogin:(void (^)(ENetError))completion
+{
+    NSBlockOperation* operation = [[NSBlockOperation alloc] init];
+    __weak NSBlockOperation* weakOperation = operation;
+    [operation addExecutionBlock:^{
+        NetCliAuthAutoReconnectEnable(false);
+
+        if (!NetCliAuthQueryConnected())
+            NetCommConnect();
+        NetCommAuthenticate(nullptr);
+
+        while (!NetCommIsLoginComplete()) {
+            if (weakOperation.cancelled) {
+                return;
+            }
+            NetCommUpdate();
+        }
+
+        ENetError result = NetCommGetAuthResult();
+        [NSOperationQueue.mainQueue addOperationWithBlock:^{
+            completion(result);
+        }];
+    }];
+    [_loginQueue addOperation:operation];
+}
+
+@end
+
+@implementation PLSLoginParameters
+
+- (id)init
+{
+    self = [super init];
+    [self load];
+    return self;
+}
+
+- (void)save
+{
+    // windows segments by product name here. in since user defaults belong to this product, we
+    // don't need to do that.
+    NSString* serverName = [NSString stringWithSTString:GetServerDisplayName()];
+    NSMutableDictionary* settingsDictionary =
+        [[[NSUserDefaults standardUserDefaults] dictionaryForKey:serverName] mutableCopy];
+    if (!settingsDictionary)
+        settingsDictionary = [NSMutableDictionary dictionary];
+    [settingsDictionary setObject:self.username forKey:@"LastAccountName"];
+    [settingsDictionary setObject:[NSNumber numberWithBool:self.rememberPassword]
+                           forKey:@"RememberPassword"];
+    [[NSUserDefaults standardUserDefaults] setObject:settingsDictionary forKey:serverName];
+    [[NSUserDefaults standardUserDefaults] synchronize];
```

Comment:
Oh right - deciding when to save the password talks to user defaults which is correct, but also platform specific. That specifically could be tossed back over to the UI side if there was deeper refactoring to share code here. I don't know if it's worth baking user defaults into Plasma itself.